Analysis route

jabb1123 opened the pull request:
This adds the routes for all analysis functions, like object tracking, safety analysis and testing scripts for feature tracking.

Note: app_config was copied to traffic_cloud_utils. I believe this should be the final location of that file and therefore any reference to the other one should be deleted eventually. The current analysis code calls the app_config file located in traffic_cloud_utils.

jabb1123 committed:
ADD: Fleshed out all analysis handlers on cloud

Need to add direct paths for file paths to finish

jabb1123 committed:
ADD: Routes for analysis now in place

Need hard-code paths for file location to finalize and merge

jabb1123 committed:
ADD: Support for emailing user on completion

jabb1123 committed:
Merge branch 'master' into analysis-route

jabb1123 committed:
EDIT: switched to prebuilt emailHandler

jabb1123 committed:
FIX: imports

jabb1123 committed:
FIX: corrected call to receive arguments

jabb1123 committed:
ADD:  Testing now implemented for object and feature tracking

jabb1123 committed:
Merge branch 'master' into analysis-route

jabb1123 committed:
Now uses updated app_config calls

jabb1123 committed:
Integrated new app_config calls

jabb1123 committed:
Copied app_config into utils folder and changed calls

jabb1123 committed:
All imports now correct

runnersaw commented on app/handlers/analysis.py:
```diff
@@ -18,4 +26,76 @@ class AnalysisHandler(tornado.web.RequestHandler):
     @apiError error_message The error message to display.
     """
     def post(self):
+        identifier = self.request.body_arguments["identifier"]
```
Do we need to do error handling (i.e. checking if "identifier" exists)?

jabb1123 replied:
We were ignoring errors for now I believe.

dcelik replied:
Agreed, definitely something that needs to be done before public release. Should create an error handling branch after MVP is complete.

runnersaw commented on app/handlers/analysis.py:
```diff
@@ -18,4 +26,76 @@ class AnalysisHandler(tornado.web.RequestHandler):
+
         self.finish("Analysis")
+
+    def safetyAnalysis(self, identifier, prediction_method=None):
```
Maybe don't duplicate these functions and call the functions from the other handlers?

jabb1123 replied:
Will do

runnersaw commented on app/handlers/objectTracking.py:
```diff
@@ -17,5 +27,56 @@ class ObjectTrackingHandler(tornado.web.RequestHandler):
+
+        # creates new config file
+        prev_tracking_path = os.path.join(project_path, ".temp", "test", "test_object", "object_tracking.cfg")
+        shutil.copyfile(prev_tracking_path, tracking_path)
```
Not 100% sure if we're doing good things with files here. Why do we check for the /run folder, and then assume that /run/run_tracking exists? Why do we delete the old object tracking, and replace it?

(IMO, the change that should be make here is deleting the check for if /run exists. If it doesn't exist, that check doesn't help anyway because run_tracking won't exist).

runnersaw commented on app/handlers/objectTracking.py:
```diff
@@ -17,5 +27,56 @@ class ObjectTrackingHandler(tornado.web.RequestHandler):
+
+        db_make_objtraj(db_path)  # Make our object_trajectories db table
+
+        video.create_tracking_video(project_path, get_project_video_path(identifier))
```
We might want to remove this for sake of time, but if we want it to check object tracking for now, then keep it.

runnersaw commented on app/handlers/testConfig.py:
```diff
@@ -20,4 +31,61 @@ class TestConfigHandler(tornado.web.RequestHandler):
+        test_flag = self.request.body_arguments["test_flag"]
+
+        try:
+            frame_start = int(self.request.body_arguments["frame_start"])
```
This won't work as expected if you specify one but not the other. Maybe:

```
frame_start = 0
num_frames = 120
if "frame_start" in self.request.body_arguments:
    frame_start = self.request.body_arguments["frame_start"]
if "num_frames" in self.request.body_arguments:
    num_frames = self.request.body_arguments["num_frames"]

dcelik replied:
yeah, we should be able to just set one without setting both to defaults

runnersaw commented on app/handlers/testConfig.py:
```diff
@@ -20,4 +31,61 @@ class TestConfigHandler(tornado.web.RequestHandler):
+
+        if test_flag == "feature":
+            runConfigTestFeature(identifier, frame_start, num_frames)
+        if test_flag == "object":
```
elif? Not super important but both shouldn't exist

dcelik replied:
I think elif is clearer in this context

runnersaw commented on app/handlers/testConfig.py:
```diff
@@ -20,4 +31,61 @@ class TestConfigHandler(tornado.web.RequestHandler):
+        subprocess.call(["feature-based-tracking", tracking_path, "--tf", "--database-filename", db_path])
+        subprocess.call(["display-trajectories.py", "-i", get_project_video_path(identifier), "-d", db_path, "-o", ac.CURRENT_PROJECT_PATH + "/homography/homography.txt", "-t", "feature", "--save-images", "-f", str(frame_start), "--last-frame", str(frame_start+num_frames)])
+
+        video.move_images_to_project_dir_folder(images_folder)
```
Why are you doing this stuff with images?

runnersaw replied:
I now remember. The reason we wanted to generate images before was that our Qt was broken and couldn't play videos. I think that the best API design would be to use `video` functions to create and return a video. We'll decompose that into frames on the client side.

runnersaw commented on app/handlers/upload.py:
```diff
@@ -3,7 +3,7 @@
 import os
 import uuid
-from trafficcloud import api
+# from trafficcloud import api
```
I'd say just go for it and remove these

dcelik replied:
Remove changes to this file, as it will become obsolete when upload_routes is merged.

runnersaw reviewed (commented):
I like the move of app_config to that folder. Have you made sure to change the imports everywhere?

dcelik commented on app/handlers/objectTracking.py:
```diff
@@ -17,5 +27,56 @@ class ObjectTrackingHandler(tornado.web.RequestHandler):
+        message = "Hello,\n\tWe have finished processing your video and identifying all objects.\nThank you for your patience,\nThe Santos Team"
+        subject = "Your video has finished processing."
+
+        EmailHelper.send_email(self.request.body_arguments["email"], subject, message)
```
It would be cool if we had a single file that contained helper methods for sending each email for the analysis.

dcelik reviewed (changes requested):
Looks mostly good, just a few things to fix and some comments by Sawyer that should be resolved or discussed.

jabb1123 committed:
EDIT: Modifications to meet PR reviews

- Fixed imports
- Called functions from files, rather than copying them
- fixed try/except to if statement
- modified if-if to if-elif for performance
- changed upload to no longer call trajectory analysis
DID NOT:
- remove app_config from trafficcloud dir
  - Still has calls in api.py and pm.py
  - These other files may be able to removed

dcelik commented on app/handlers/testConfig.py:
```diff
@@ -35,17 +32,18 @@ def post(self):
+        if "frame_start" in self.request.body_arguments:
             frame_start = int(self.request.body_arguments["frame_start"])
+
+        if "num_frames" in self.request.body_arguments:
```
Good bug fix.

dcelik committed:
Update version number to reflect working state

dcelik committed:
EDIT: Version number to reflect working state

dcelik committed:
EDIT: Version number to reflect working state

dcelik committed:
EDIT: Version number to reflect working state

runnersaw commented on app/handlers/analysis.py:
```diff
@@ -1,12 +1,21 @@
+from traffic_cloud_utils.app_config import get_project_path, get_project_video_path, update_config_without_sections, get_config_without_sections
+from traffic_cloud_utils.emailHelper import EmailHelper
+from objectTracking import objectTrack
+from safetyAnalysis import safetyAnalysis
```
I don't think you can import these functions like this. These functions are methods of the ObjectTrackingHandler class, so they aren't in the top level available for import.

jabb1123 committed:
FIX: Minor fixes to make funcional

- import changes
- Changed how email handler is constructed
- fixed body calls

jabb1123 committed:
Merge branch 'master' into analysis-route

jabb1123 committed:
FIX: Error in receiving data

jabb1123 committed:
Fix: Spelling

jabb1123 committed:
More import fixes

jabb1123 committed:
RM: references to app_config

jabb1123 committed:
EDIT: Changed int to string

jabb1123 committed:
Testing debug

jabb1123 committed:
FIX: image moving now working correctly

jabb1123 committed:
Fix call to old move function

jabb1123 committed:
FIX: Call to handler

jabb1123 committed:
EDIT: added self

jabb1123 committed:
ADD: Import for plotting

jabb1123 committed:
Corrected import

jabb1123 committed:
Fixed more imports

jabb1123 committed:
ADD: Output to tell that Safety Analysis is being run serverside

jabb1123 committed:
EDIT: Finalize changes to analysis routes

dcelik commented on app/handlers/analysis.py:
```diff
@@ -18,4 +26,17 @@ class AnalysisHandler(tornado.web.RequestHandler):
+        identifier = self.get_body_argument("identifier")
+
+        ObjectTrackingHandler.handler(identifier)
+        SafetyAnalysisHandler.handler(identifier)
```
What if the first handler fails but the second one succeeds? A check should be done in the middle, however unlikely the occurance may be.

dcelik commented on app/handlers/analysis.py:
```diff
@@ -18,4 +26,17 @@ class AnalysisHandler(tornado.web.RequestHandler):
+        ObjectTrackingHandler.handler(identifier)
+        SafetyAnalysisHandler.handler(identifier)
+
+        if status_code == 200:
```
where is status_code created?

jabb1123 committed:
FIX: Meet all of Deniz Concerns on PR
